README.md
# promise-debounce-pool

Maintains a pool of leading edge debounced promises.

This is useful if you want to limit MULTIPLE expensive (but independant) operations (such as api calls) and only 
have one occur at any one time (synchronously)

You `set` the promise resolver you want to debounce and when you `get` it, the pool actually handles creating the promise, and then
only when it needs to. If one is already pending for that key, it returns that.

Calls between different promise resolvers (different keys) in the pool are chained synchronously (one after the other).
If a call to get a promise from the pool occurs while another promise (one with a different key) is 'pending'
then the second call is waits for the other 'pending' promise to resolve/reject before starting.

The chained promises do not react to or interfere with the previous promise in any way. 
i.e. The chained promise does not fail if the previous promise fails, nor does the returned data bleed into the 
second call

## Example Usage

[![Try promise-debounce-pool on RunKit](https://badge.runkitcdn.com/promise-debounce-pool.svg)](https://npm.runkit.com/promise-debounce-pool)

```js
var PromisePool = require("promise-debounce-pool");

// create a promise resolver function for
// each of your expensive operations

var expensiveFooOperation = function(resolve, reject) {
   setTimeout(function() { resolve('foo data ' + Math.random()) }, 1000);
};

var expensiveBarOperation = function(resolve, reject) {
   setTimeout(function() { resolve('bar data ' + Math.random()) }, 1000);
};

// create an instance of the pool
var promisePool = new PromisePool();

// associate each promise resolver to a key in the pool
// Note: the functions are not invoked at this point
promisePool.set('foo', expensiveFooOperation);
promisePool.set('bar', expensiveBarOperation);

// make multiple calls

var fooCall1 = promisePool.get('foo') //starts a new foo promise
    .then(function(fooData) {
         console.log('fooCall1: ' + fooData);
    });

var fooCall2 = promisePool.get('foo') //returns the same pending foo promise (fooCall1)
    .then(function(fooData) {
        console.log('fooCall2: ' + fooData); //same data from fooCall1
    });

var barCall1 = promisePool.get('bar') //waits until fooCall1 promise resolved/rejected then creates a new bar promise
    .then(function(barData) {
        console.log('barCall1: ' + barData);
    });

var barCall2 = promisePool.get('bar') //returns the same pending bar promise (barCall1)
    .then(function(barData) {
        console.log('barCall2: ' + barData); // same data from barCall1
    });

setTimeout(function() {

 var fooCall3 = promisePool.get('foo') //starts a new foo promise (since previous fooCall1 promise has now resolved)
    .then(function(fooData) {
        console.log('fooCall3: ' + fooData); // data from new fooCall3 promise
    });


 var barCall3 = promisePool.get('bar') //returns the same fooCall1 promise from before, its still pending
    .then(function(barData) {
        console.log('barCall3: ' + barData); // same data from barCall1
    });

}, 1500);


//RESULTS IN:
// expensiveFooOperation called twice. 
//   Once for fooCall1 & fooCall2 because fooCall1 was still pending whn fooCall2 occurred
//   Then again for fooCall3 because fooCall1 had resolved by the time fooCall3 occurred. 
//
// expensiveBarOperation called only once
//   This occurred after the first call to expensiveFooOperation completed. 
//   All three bar calls shared the same promise becuase they all occurred before barCall1 had resolved.
```

## Tests:

```
npm install
npm test
```

## API:

##### set
` promisePool.set(key, fn)` 

Set the promise resolver

`key` used to identify the promise when getting/setting \
`fn` resolver function used when creating a new promise for this key. e.g.\
```js
promisePool.set('someKey', function(resolve, reject) {
    //...
    resolve();
})
```

`fn` can optionally be curried if you want to pass other arguments to it when it is invoked.
In this case the function passed to `set` must return the resolver function.

```js
promisePool.set('someKey', function yummyCurry(curriedArg1, curriedArg2) {
    return function resolver(resolve, reject) {
        //...
        resolve();
    };
});

promisePool.get('someKey', 'curriedValue1', 'curriedValue2')
    .then(
        //...
     );
```



##### get
` var promise = promisePool.get(key, [curriedArg], [...])`

Get a promise from the pool

`key` used to identify the promise when getting/setting \
Optionally you can pass multiple `curriedArg` which will be passed on to the curried promise resolver.
If you pass more arguments to get then you must curry the resolver when.

If you call `get` with a given `key`, it returns you a `promise` based on the following:
1) is there already a pending promise for this `key`, if so return it
2) is there currently a pending promise from another `key`, if so create a new promise but chain
   onto the pending promise so it happens synchronously
3) create a new promise for this key

**Note**: since multiple calls share pending promises, if you make multiple calls with different curried args
while the first is still pending, then the other calls will still share the first call's promise and curried args. e.g.

```js
promisePool.set('someKey', function yummyCurry(curriedArg1, curriedArg2) {
    return function resolver(resolve, reject) {
        setTimeout(function() { resolve() }, 1000);
    };
});

var call1 = promisePool.get('someKey', 'curriedValue1', 'curriedValue2')
    .then(
        //...
     );

var call2 = promisePool.get('someKey', 'curriedValue3', 'curriedValue4')
    .then(
        // ... (data from call1 with call1's curried args, since call1 was stil pending)
     );
```

##### isSet
` var isSet = promisePool.isSet(key)`

Is a promise resolver set in the pool

`key` used to identify the promise in the pool \
`isSet` returns a boolean

## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details


CHANGES.md
## v1.4.0 (2017-09-28):

- Added UMD support so that it can be used in more than just node
- Added minified version and source map
- Added runkit example
- Updated README example
- Added whitelist files to publish to npm in package.json
- Switched to MIT license
- Added CHANGES.md


spec/promise-pool-spec.js
var PromisePool = require('../promise-debounce-pool');

describe("PromisePool", function() {
    describe("call to get", function() {

        it('returns a resolved promise with data from the resolver', function(done) {
            var expectedData = 'data returned from promise';
            var promisePool = new PromisePool();
            promisePool.set('promise_key_here', function(resolve) {
                resolve(expectedData);
            });
            promisePool.get('promise_key_here')
                .then(function(actualData) {
                    expect(actualData).toBe(expectedData);
                    done();
                });
        });

        it('returns a rejected promise with error from the resolver', function(done) {
            var expectedError = 'error returned from promise';
            var promisePool = new PromisePool();
            promisePool.set('promise_key_here', function(resolve, reject) {
                reject(expectedError);
            });
            promisePool.get('promise_key_here')
                .then(function () {
                    throw new Error('promise did not reject');
                })
                .catch(function(actualError) {
                    expect(actualError).toBe(expectedError);
                    done();
                });
        });

        it("returns rejected promise if no resolver set", function (done) {

            var promisePool = new PromisePool();
            promisePool.get('promise_key_here')
                .then(function () {
                    throw new Error('promise did not reject');
                })
                .catch(function (actualError) {
                    expect(actualError.name).toBe('TypeError');
                    expect(actualError.message).toBe(
                        'Promise resolver undefined is not a function for key promise_key_here'
                    );
                    done();
                });
        });

        it('returns rejected promise if resolver is not a function', function (done) {
            var promisePool = new PromisePool();
            promisePool.set('promise_key_here', 'not_a_function');
            promisePool.get('promise_key_here')
                .then(function () {
                    throw new Error('promise did not reject');
                })
                .catch(function (actualError) {
                    expect(actualError.name).toBe('TypeError');
                    expect(actualError.message).toBe(
                        'Promise resolver not_a_function is not a function for key promise_key_here'
                    );
                    done();
                });
        });

        it('calls the resolver', function(done) {
            var promisePool = new PromisePool();
            var observer = {
                resolver: function(resolve) {
                    resolve();
                }
            };
            var resolverSpy = spyOn(observer, 'resolver').and.callThrough();
            promisePool.set('promise_key_here', resolverSpy);
            promisePool.get('promise_key_here')
                .then(function() {
                    expect(resolverSpy).toHaveBeenCalledTimes(1);
                    done();
                });
        });
    });

    describe('multiple calls to get while promise still pending', function () {

        it('returns same resolved promise with data from the resolver', function(done) {
            var expectedData = 'data returned from promise';
            var promisePool = new PromisePool();

            jasmine.clock().install();
            promisePool.set('promise_key_here', function(resolve) {
                setTimeout(function() { resolve(expectedData) }, 100);
            });
            promisePool.get('promise_key_here')
                .then(function(actualData) {
                    expect(actualData).toBe(expectedData);
                });

            promisePool.get('promise_key_here')
                .then(function(actualData) {
                    expect(actualData).toBe(expectedData);
                });

            promisePool.get('promise_key_here')
                .then(function(actualData) {
                    expect(actualData).toBe(expectedData);
                    done();
                });

            jasmine.clock().tick(101);
            jasmine.clock().uninstall();
        });

        it('only calls the resolver once', function(done) {
            var expectedData = 'data returned from promise';
            var promisePool = new PromisePool();
            var observer = {
                resolver: function(resolve) {
                    setTimeout(function() { resolve(expectedData) }, 100);
                }
            };
            var resolverSpy = spyOn(observer, 'resolver').and.callThrough();

            jasmine.clock().install();
            promisePool.set('promise_key_here', resolverSpy);
            promisePool.get('promise_key_here');
            promisePool.get('promise_key_here');

            promisePool.get('promise_key_here')
                .then(function() {
                    expect(resolverSpy).toHaveBeenCalledTimes(1);
                    done();
                });

            jasmine.clock().tick(101);
            jasmine.clock().uninstall();
        });

        it('returns same rejected promise with error from the resolver', function(done) {
            var expectedError = 'error returned from promise';
            var promisePool = new PromisePool();

            jasmine.clock().install();
            promisePool.set('promise_key_here', function(resolve, reject) {
                setTimeout(function() { reject(expectedError); }, 100);
            });

            promisePool.get('promise_key_here')
                .then(function () {
                    throw new Error('promise did not reject');
                })
                .catch(function(actualError) {
                    expect(actualError).toBe(expectedError);
                });

            promisePool.get('promise_key_here')
                .then(function () {
                    throw new Error('promise did not reject');
                })
                .catch(function(actualError) {
                    expect(actualError).toBe(expectedError);
                });

            promisePool.get('promise_key_here')
                .then(function () {
                    throw new Error('promise did not reject');
                })
                .catch(function(actualError) {
                    expect(actualError).toBe(expectedError);
                    done();
                });

            jasmine.clock().tick(101);
            jasmine.clock().uninstall();
        });
    });

    describe('multiple calls to get after promise resolved', function () {

        it('returns new resolved promises with data from the resolver', function(done) {
            var increment = 0;
            var promisePool = new PromisePool();
            var observer = {
                resolver: function(resolve) {
                    resolve(++increment);
                }
            };
            var resolverSpy = spyOn(observer, 'resolver').and.callThrough();

            promisePool.set('promise_key_here', resolverSpy);
            promisePool.get('promise_key_here')
                .then(function(actualData) {

                    expect(actualData).toBe(1);
                    expect(resolverSpy).toHaveBeenCalledTimes(1);

                    promisePool.get('promise_key_here')
                        .then(function(actualData) {

                            expect(actualData).toBe(2);
                            expect(resolverSpy).toHaveBeenCalledTimes(2);

                            promisePool.get('promise_key_here')
                                .then(function (actualData) {
                                    expect(actualData).toBe(3);
                                    expect(resolverSpy).toHaveBeenCalledTimes(3);
                                    done();
                                });
                        });
                });
        });

        it('returns new resolved promises with data from the resolver', function(done) {
            var increment = 0;
            var promisePool = new PromisePool();
            var observer = {
                resolver: function(resolve) {
                    resolve(++increment);
                }
            };
            var resolverSpy = spyOn(observer, 'resolver').and.callThrough();

            promisePool.set('promise_key_here', resolverSpy);
            promisePool.get('promise_key_here')
                .then(function(actualData) {
                    expect(actualData).toBe(1);
                    expect(resolverSpy).toHaveBeenCalledTimes(1);
                });

            setTimeout(function() {
                promisePool.get('promise_key_here')
                    .then(function (actualData) {
                        expect(actualData).toBe(2);
                        expect(resolverSpy).toHaveBeenCalledTimes(2);
                    });
            }, 10);

            setTimeout(function() {
                promisePool.get('promise_key_here')
                    .then(function (actualData) {
                        expect(actualData).toBe(3);
                        expect(resolverSpy).toHaveBeenCalledTimes(3);
                        done();
                    });
            }, 20);
        });
    });

    describe('multiple calls to get with different keys after promise resolved', function () {

        it('returns new resolved promises with data from the resolver', function (done) {

            var expectedDataFromA = 'data from promise A';
            var expectedDataFromB = 'data from promise B';
            var promisePool = new PromisePool();

            jasmine.clock().install();

            promisePool.set('promiseA', function (resolve) {
                setTimeout(function () {
                    resolve(expectedDataFromA);
                }, 100);
            });

            promisePool.set('promiseB', function (resolve) {
                setTimeout(function () {
                    resolve(expectedDataFromB);
                }, 100);
            });

            promisePool.get('promiseA')
                .then(function (actualDataFromA) {
                    expect(actualDataFromA).toBe(expectedDataFromA);
                });

            jasmine.clock().tick(101);

            promisePool.get('promiseB')
                .then(function (actualDataFromB) {
                    expect(actualDataFromB).toBe(expectedDataFromB);
                    done();
                });

            jasmine.clock().tick(101);
            jasmine.clock().uninstall();
        });
    });

    describe('multiple calls to get with different keys while still pending', function () {

        it('chains second promise after the first', function (done) {
            var promisePool = new PromisePool();
            var observer = {
                resolverA: function (resolve) {
                    setTimeout(function () {
                        resolve();
                    }, 100);
                },
                resolverB: function (resolve) {
                    setTimeout(function () {
                        resolve();
                    }, 100);
                }
            };
            var resolverSpyA = spyOn(observer, 'resolverA').and.callThrough();
            var resolverSpyB = spyOn(observer, 'resolverB').and.callThrough();

            jasmine.clock().install();

            promisePool.set('promiseA', resolverSpyA);
            promisePool.set('promiseB', resolverSpyB);

            promisePool.get('promiseA')
                .then(function () {
                    expect(resolverSpyA).toHaveBeenCalledTimes(1);
                    expect(resolverSpyB).toHaveBeenCalledTimes(0);
                    done();
                });

            promisePool.get('promiseB')
                .then(function () {
                    expect(resolverSpyA).toHaveBeenCalledTimes(1);
                    expect(resolverSpyB).toHaveBeenCalledTimes(1);
                    done();
                });

            jasmine.clock().tick(101);
            jasmine.clock().uninstall();
        });

        it('chains third promise after the second, etc...', function (done) {
            var promisePool = new PromisePool();
            var observer = {
                resolverA: function (resolve) {
                    setTimeout(function () {
                        resolve();
                    }, 100);
                },
                resolverB: function (resolve) {
                    setTimeout(function () {
                        resolve();
                    }, 100);
                },
                resolverC: function (resolve) {
                    setTimeout(function () {
                        resolve();
                    }, 100);
                }
            };
            var resolverSpyA = spyOn(observer, 'resolverA').and.callThrough();
            var resolverSpyB = spyOn(observer, 'resolverB').and.callThrough();
            var resolverSpyC = spyOn(observer, 'resolverC').and.callThrough();

            jasmine.clock().install();

            promisePool.set('promiseA', resolverSpyA);
            promisePool.set('promiseB', resolverSpyB);
            promisePool.set('promiseC', resolverSpyC);

            promisePool.get('promiseA')
                .then(function () {
                    expect(resolverSpyA).toHaveBeenCalledTimes(1);
                    expect(resolverSpyB).toHaveBeenCalledTimes(0);
                    expect(resolverSpyC).toHaveBeenCalledTimes(0);
                });

            promisePool.get('promiseB')
                .then(function () {
                    expect(resolverSpyA).toHaveBeenCalledTimes(1);
                    expect(resolverSpyB).toHaveBeenCalledTimes(1);
                    expect(resolverSpyC).toHaveBeenCalledTimes(0);
                });


            promisePool.get('promiseC')
                .then(function () {
                    expect(resolverSpyA).toHaveBeenCalledTimes(1);
                    expect(resolverSpyB).toHaveBeenCalledTimes(1);
                    expect(resolverSpyC).toHaveBeenCalledTimes(1);
                    done();
                });

            jasmine.clock().tick(101);
            jasmine.clock().uninstall();
        });

        it('shares same promise even between other calls', function (done) {
            var promisePool = new PromisePool();
            var observer = {
                resolverA: function (resolve) {
                    setTimeout(function () {
                        resolve();
                    }, 100);
                },
                resolverB: function (resolve) {
                    setTimeout(function () {
                        resolve();
                    }, 100);
                }
            };
            var resolverSpyA = spyOn(observer, 'resolverA').and.callThrough();
            var resolverSpyB = spyOn(observer, 'resolverB').and.callThrough();

            jasmine.clock().install();

            promisePool.set('promiseA', resolverSpyA);
            promisePool.set('promiseB', resolverSpyB);

            promisePool.get('promiseA')
                .then(function () {
                    expect(resolverSpyA).toHaveBeenCalledTimes(1);
                    expect(resolverSpyB).toHaveBeenCalledTimes(0);
                });

            promisePool.get('promiseB')
                .then(function () {
                    expect(resolverSpyA).toHaveBeenCalledTimes(1);
                    expect(resolverSpyB).toHaveBeenCalledTimes(1);
                });

            promisePool.get('promiseA')
                .then(function () {
                    expect(resolverSpyA).toHaveBeenCalledTimes(1);
                    expect(resolverSpyB).toHaveBeenCalledTimes(1);
                    done();
                });

            jasmine.clock().tick(101);
            jasmine.clock().uninstall();
        });
    });

    describe('multiple chained calls', function () {

        it('maintain separate resolve, reject states and data', function (done) {
            var expectedError = 'Error thrown by promiseA';
            var expectedData = 'Data returned by promiseB';
            var promisePool = new PromisePool();
            var observer = {
                resolverA: function (resolve, reject) {
                    setTimeout(function () {
                        reject(expectedError);
                    }, 100);
                },
                resolverB: function (resolve) {
                    setTimeout(function () {
                        resolve(expectedData);
                    }, 100);
                }
            };
            var resolverSpyA = spyOn(observer, 'resolverA').and.callThrough();
            var resolverSpyB = spyOn(observer, 'resolverB').and.callThrough();

            jasmine.clock().install();

            promisePool.set('promiseA', resolverSpyA);
            promisePool.set('promiseB', resolverSpyB);

            promisePool.get('promiseA')
                .then(function () {
                    throw new Error('promise did not reject');
                })
                .catch(function(expectedError) {
                    expect(expectedError).toBe(expectedError);
                });

            promisePool.get('promiseB')
                .then(function (actualData) {
                    expect(actualData).toBe(expectedData);
                    done();
                });

            jasmine.clock().tick(101);

            promisePool.get('promiseA')
                .then(function () {
                    throw new Error('promise did not reject');
                })
                .catch(function(actualError) {
                    expect(actualError).toBe(expectedError);
                });

            jasmine.clock().tick(101);
            jasmine.clock().uninstall();
        });
    });

    describe('Currying resolver', function () {
        it('passes arguments from get on to curried resolver', function (done) {

            var expectedData = 'foo data with curried arg 1 and curried arg 2';

            var promisePool = new PromisePool();
            var observer = {
                yummyCurry: function(curriedArg1, curriedArg2) {
                    return function resolver(resolve) {
                        resolve('foo data ' + curriedArg1 + ' ' + curriedArg2);
                    };
                }
            };

            var currySpy = spyOn(observer, 'yummyCurry').and.callThrough();

            promisePool.set('foo', currySpy);

            promisePool.get('foo', 'with curried arg 1', 'and curried arg 2')
                .then(function (actualData) {
                    expect(currySpy).toHaveBeenCalledWith('with curried arg 1', 'and curried arg 2');
                    expect(actualData).toBe(expectedData);
                    done();
                });
        });

        it('returns a rejected promise if resolver is not curried', function (done) {
            var promisePool = new PromisePool();
            promisePool.set('promise_key_here', function() {
                return 'not_a_resolver_function';
            });
            promisePool.get('promise_key_here', 'some_curried_arg')
                .then(function () {
                    throw new Error('promise did not reject');
                })
                .catch(function (actualError) {
                    expect(actualError.name).toBe('TypeError');
                    expect(actualError.message).toBe(
                        'Curried promise resolver not_a_resolver_function is not a function for key promise_key_here'
                    );
                    done();
                });
        });
    });

    describe('Currying resolver, multiple calls while pending', function () {
        it('share the same promise with data from the first call', function(done) {
            var expectedDataFromCall1 = 'data returned from promise with data from call 1';
            var promisePool = new PromisePool();

            jasmine.clock().install();
            promisePool.set('promise_key_here', function yummyCurry(curriedArg) {
                return function resolver(resolve) {
                    resolve('data returned from promise ' + curriedArg)
                }
            });
            promisePool.get('promise_key_here', 'with data from call 1')
                .then(function(actualData) {
                    expect(actualData).toBe(expectedDataFromCall1);
                });

            promisePool.get('promise_key_here', 'with data from call 2')
                .then(function(actualData) {
                    expect(actualData).toBe(expectedDataFromCall1);
                });

            promisePool.get('promise_key_here', 'with data from call 3')
                .then(function(actualData) {
                    expect(actualData).toBe(expectedDataFromCall1);
                    done();
                });

            jasmine.clock().tick(101);
            jasmine.clock().uninstall();
        });
    });
});


src/pool.js
/**
 * PromisePool
 *
 * maintains a pool of promises that are 'pending' so that multiple calls return that same promise.
 *
 * The same pending promise is returned for all calls to get it from the pool until it resolves/rejects,
 * at which point on the next call, it creates a new promise
 *
 * This is useful if you want to limit the number of identical api requests in quick succession.
 *
 * Calls between different promises in the pool are performed synchronously (one after the other).
 * If a call to get a promise from the pool occurs while different promise (one with a different key) is 'pending'
 * the second call is waits for the other 'pending' promise to resolve/reject before starting.
 * (it chains the second call using .then off of the previous call)
 * The chained promises do not react to or interfere with the previous promise in any way. any data or errors are
 * passed straight through. The chained promise does not fail if the previous promise fails
 *
 * @example:
 *
 *    var promisePool = new PromisePool();
 *
 *    promisePool.set('foo', function(resolve, reject) {
 *          setTimeout(function() { resolve('foo data') }, 1000);
 *    });
 *
 *    promisePool.set('bar', function(resolve, reject) {
 *          setTimeout(function() { resolve('bar data') }, 1000);
 *    });
 *
 *    var call1 = promisePool.get('foo'); //starts a new foo promise
 *    var call2 = promisePool.get('foo'); //returns the same pending foo promise
 *    var call3 = promisePool.get('bar'); //waits until foo promise resolved/rejected then creates a new bar promise
 *    var call4 = promisePool.get('bar'); //returns the waiting foo promise
 *
 *    setTimeout(function() {
 *      var call5 = promisePool.get('foo'); //starts a new foo promise (since previous foo promise has now resolved)
 *      var call6 = promisePool.get('bar'); //returns the same bar promise from before, now pending
 *    }, 1000);
 *
 *    //RESULTS IN:
 *    //Two foo promises created, with the second foo call sharing the first
 *    //One bar promise created with the second and third bar calls sharing the first
 *
 * @constructor
 */
function PromisePool() {
    this.promise = {};
    this.resolver = {};
}

/**
 * Is a promise set in the pool
 *
 * @param {*} key - used to identify the promise when getting/setting
 * @returns {boolean}
 */
PromisePool.prototype.isSet = function(key) {
    return typeof this.resolver[key] !== 'undefined';
};

/**
 * Set the promise resolver
 *
 * @param {*} key - used to identify the promise when getting/setting
 * @param {function} resolver - resolver function used when creating a new promise for this key
 */
PromisePool.prototype.set = function(key, resolver) {
    this.resolver[key] = resolver;
};

/**
 * Get a promise from the pool
 *
 * Logic:
 * if you call get with a given key, it returns you a promise based on the following:
 * 1) is there already a pending promise for this key, if so return it
 * 2) is there currently a pending promise from another key, if so create a new promise but chain
 *    onto the pending promise so it happens synchronously
 * 3) create a new promise for this key
 *
 * @param {*} key - used to identify the promise when getting/setting
 * @returns {Promise}
 */
PromisePool.prototype.get = function(key) {

    var _this = this;

    if (this.promise[key]) {
        return this.promise[key];
    }

    if (this.pending) {
        this.promise[key] = this.pending.then(createPromise, createPromise);
        this.pending = this.promise[key];
        return this.promise[key];
    }

    this.promise[key] = createPromise();
    this.pending = this.promise[key];
    return this.promise[key];

    function createPromise() {
        var resolver = _this.resolver[key];
        if (typeof resolver !== 'function') {
            return exports.Promise.reject(
                new TypeError('Promise resolver ' + resolver + ' is not a function for key ' + key)
            );
        }
        return new exports.Promise(resolver)
            .then(function(success) {
                delete _this.promise[key];
                delete _this.pending;
                return success;
            }, function(error) {
                delete _this.promise[key];
                delete _this.pending;
                throw error;
            });
    }
};

exports.PromisePool = PromisePool;

promise-pool.js

function PromisePool() {
    this.promise = {};
    this.resolver = {};
}

PromisePool.prototype.isSet = function(key) {
    return typeof this.resolver[key] !== 'undefined';
};

PromisePool.prototype.set = function(key, resolver) {
    this.resolver[key] = resolver;
};

PromisePool.prototype.get = function(key) {

    var _this = this;

    if (this.promise[key]) {
        return this.promise[key];
    }

    if (this.pending) {
        this.promise[key] = this.pending.then(createPromise, createPromise);
        this.pending = this.promise[key];
        return this.promise[key];
    }

    this.promise[key] = createPromise();
    this.pending = this.promise[key];
    return this.promise[key];

    function createPromise() {
        var resolver = _this.resolver[key];
        if (typeof resolver !== 'function') {
            return Promise.reject(
                new TypeError('Promise resolver ' + resolver + ' is not a function for key ' + key)
            );
        }
        return new Promise(resolver)
            .then(function(success) {
                delete _this.promise[key];
                delete _this.pending;
                return success;
            }, function(error) {
                delete _this.promise[key];
                delete _this.pending;
                throw error;
            });
    }
};

module.exports = PromisePool;


example.runkit.js
var PromisePool = require("promise-debounce-pool");

// create a promise resolver function for
// each of your expensive operations

var expensiveFooOperation = function(resolve, reject) {
    setTimeout(function() { resolve('foo data ' + Math.random()) }, 1000);
};

var expensiveBarOperation = function(resolve, reject) {
    setTimeout(function() { resolve('bar data ' + Math.random()) }, 1000);
};

// create an instance of the pool
var promisePool = new PromisePool();

// associate each promise resolver to a key in the pool
// Note: the functions are not invoked at this point
promisePool.set('foo', expensiveFooOperation);
promisePool.set('bar', expensiveBarOperation);

// make multiple calls

var fooCall1 = promisePool.get('foo') //starts a new foo promise
    .then(function(fooData) {
        console.log('fooCall1: ' + fooData);
    });

var fooCall2 = promisePool.get('foo') //returns the same pending foo promise (fooCall1)
    .then(function(fooData) {
        console.log('fooCall2: ' + fooData); //same data from fooCall1
    });

var barCall1 = promisePool.get('bar') //waits until fooCall1 promise resolved/rejected then creates a new bar promise
    .then(function(barData) {
        console.log('barCall1: ' + barData);
    });

var barCall2 = promisePool.get('bar') //returns the same pending bar promise (barCall1)
    .then(function(barData) {
        console.log('barCall2: ' + barData); // same data from barCall1
    });

setTimeout(function() {

    var fooCall3 = promisePool.get('foo') //starts a new foo promise (since previous fooCall1 promise has now resolved)
        .then(function(fooData) {
            console.log('fooCall3: ' + fooData); // data from new fooCall3 promise
        });


    var barCall3 = promisePool.get('bar') //returns the same fooCall1 promise from before, its still pending
        .then(function(barData) {
            console.log('barCall3: ' + barData); // same data from barCall1
        });

}, 1500);
